Force :attachment as content disposition for some content types

rosa opened the pull request:
### Summary

Currently, unless a value for `disposition` is passed to `ActiveStorage::Blob#service_url` we use `inline` for everything. However, some content types shouldn't be rendered inline by default. For example:
- Sending `.ai` files as application/postscript to Safari opens them in a blank, grey screen, and downloading `.ai` as `application/postscript` files in Safari appends `.ps` to the extension.
- Sending HTML, SVG, XML and SWF files as binary closes XSS vulnerabilities. Even if these are mitigated by serving the blobs from a different domain from your application, having an HTML file rendered by the browser instead of downloaded opens the door for more realistic phishing attacks.
- Sending JS files as binary avoids `InvalidCrossOriginRequest` without compromising security.

This pull request modifies this behaviour to force `:attachment` as disposition for specific content types like the above, configured in `config.active_storage.content_types_to_serve_as_binary`.

### Other Information

I have included a default list of content types for this new config option, but it could be left empty if you think that makes more sense, and let users decide what they want to be treated as attachments. I think the default list makes sense, though. 

r? @georgeclaghorn

rosa committed:
Force content disposition to attachment for specific content types

In this way we avoid HTML, XML, SVG and other files that can be rendered
by the browser to be served inline by default. Depending on the origin
from where these files are served, this might lead to XSS
vulnerabilities, and in the best case, to more realistic phishing
attacks and open redirects.

We force it rather than falling back to it when other disposition is not
provided. Otherwise it would be possible for someone to force inline
just by passing `disposition=inline` in the URL.

The list of content types to be served as attachments is configurable.

rosa commented on activestorage/lib/active_storage/engine.rb:
```diff
@@ -18,6 +18,14 @@ class Engine < Rails::Engine # :nodoc:
     config.active_storage.analyzers = [ ActiveStorage::Analyzer::ImageAnalyzer, ActiveStorage::Analyzer::VideoAnalyzer ]
     config.active_storage.paths = ActiveSupport::OrderedOptions.new
     config.active_storage.variable_content_types = [ "image/png", "image/gif", "image/jpg", "image/jpeg", "image/vnd.adobe.photoshop" ]
+    config.active_storage.content_types_to_render_as_binary = [ "text/html",
```
Perhaps a better name for this would e `content_types_to_render_as_attachment`.

georgeclaghorn replied:
I’m fine with either `binary` or `attachment`, but `render` doesn’t seem like the right word to me. How about `content_types_to_serve_as_binary`?

rosa commented on activestorage/app/models/active_storage/blob.rb:
```diff
@@ -143,6 +143,16 @@ def variable?
+    else
+      :inline
+    end
+  end
```
I had put this method as `private` before, then I thought it could make sense to expose it, for example if we were to check this for blob's variants in the future (although none of the variable content types makes sense as attachment right now).

georgeclaghorn replied:
I think we should keep it private for now. If we need to call it elsewhere, we can make it public at that time.

Let’s also rename it to `default_disposition` to clarify that it’s a fallback.

georgeclaghorn commented on activestorage/lib/active_storage/engine.rb:
```diff
@@ -18,6 +18,14 @@ class Engine < Rails::Engine # :nodoc:
+                                                                "application/x-shockwave-flash",
+                                                                "text/xml",
+                                                                "application/xml",
+                                                                "application/xhtml+xml" ]
```
Style nit: indent the array elements one level under the assignment.

```ruby
config.active_storage.content_types_to_render_as_binary = [
  "text/html",
  "text/javascript",
  # ...
]
```

rosa replied:
Sure thing! The truth is, I prefer one level under the assignment too! But I wasn't sure which one would be preferred, so I searched in the codebase and found this example: 

https://github.com/rails/rails/blob/5a5014688873f1d6e1b66075eea8a4356b5a4d07/railties/lib/rails/tasks/tmp.rake#L7-L10

And did the same. But it's true in that example the indentation is much smaller, so it's not really applicable directly 😅

georgeclaghorn reviewed (changes requested):
Nice work, Rosa!

rosa commented on activestorage/app/models/active_storage/blob.rb:
```diff
@@ -201,8 +201,8 @@ def representable?
   # it allows permanent URLs that redirect to the +service_url+ to be cached in the view.
-  def service_url(expires_in: service.url_expires_in, disposition: "inline")
-    service.url key, expires_in: expires_in, disposition: disposition, filename: filename, content_type: content_type
+  def service_url(expires_in: service.url_expires_in, disposition: :inline)
```
I've changed this from `"inline"` to `:inline` because I've seen it used as symbol in other places like:
https://github.com/rails/rails/blob/5a5014688873f1d6e1b66075eea8a4356b5a4d07/activestorage/app/models/active_storage/variant.rb#L70-L72

https://github.com/rails/rails/blob/5a5014688873f1d6e1b66075eea8a4356b5a4d07/activestorage/lib/active_storage/service.rb#L91-L93

rosa commented:
@georgeclaghorn, thanks a lot for your review! I have made the changes you suggested, and also, as you very correctly pointed out, I changed the code to _force_ the content disposition to `attachment` instead of defaulting to it. Otherwise this wouldn't solve the security concerns, as an attacker could simply add `disposition=inline` to the URL to get the same effect as now. I have also updated the pull request description to match this new behaviour.

I was thinking... should this be included in the `CHANGELOG`?

georgeclaghorn commented on activestorage/app/models/active_storage/blob.rb:
```diff
@@ -201,8 +201,8 @@ def representable?
-  def service_url(expires_in: service.url_expires_in, disposition: "inline")
-    service.url key, expires_in: expires_in, disposition: disposition, filename: filename, content_type: content_type
+  def service_url(expires_in: service.url_expires_in, disposition: :inline)
+    service.url key, expires_in: expires_in, disposition: forced_disposition || disposition, filename: filename, content_type: content_type
```
What do you think about hoisting the conditional into this method, since we’re not returning more than one override from `forced_disposition`?

```ruby
service.url key, expires_in: expires_in, disposition: forcibly_serve_as_binary? ? :attachment : disposition, filename: filename, content_type: content_type
```

rosa replied:
Sounds good to me, it looks long but I think it's clearer 👍

georgeclaghorn commented:
> I was thinking... should this be included in the CHANGELOG?

Yes please!

georgeclaghorn reviewed (approved):
:heart:

Can you squash your commits into one?

rosa commented:
Of course!
